src/icelake/implementation.cpp

```diff
@@ -1040,21 +1042,65 @@ simdutf_warn_unused size_t implementation::count_utf16be(const char16_t * input,
 
 
 simdutf_warn_unused size_t implementation::count_utf8(const char * input, size_t length) const noexcept {
-  const char* end = length >= 64 ? input + length - 64 : nullptr;
-  const char* ptr = input;
-
+  const uint8_t *str = reinterpret_cast<const uint8_t *>(input);
+  size_t answer =  length / sizeof(__m512i) * sizeof(__m512i); // Number of 512-bit chunks that fits into the length.
+  size_t i = 0;
+  __m512i unrolled_popcount{0}; 
   const __m512i continuation = _mm512_set1_epi8(char(0b10111111));
 
-  size_t count{0};
+  while (i + sizeof(__m512i) <= length) {
+    size_t iterations = (length - i) / sizeof(__m512i);
+    if (iterations > UINT64_MAX) { 
+      iterations = UINT64_MAX;
+    }
+    size_t max_i = i + iterations * sizeof(__m512i) - sizeof(__m512i);
+    for (; i + 8*sizeof(__m512i) <= max_i; i += 8*sizeof(__m512i)) {
+        __m512i input1 = _mm512_loadu_si512((const __m512i *)(str + i));
+        __m512i input2 = _mm512_loadu_si512((const __m512i *)(str + i + sizeof(__m512i)));
+        __m512i input3 = _mm512_loadu_si512((const __m512i *)(str + i + 2*sizeof(__m512i)));
+        __m512i input4 = _mm512_loadu_si512((const __m512i *)(str + i + 3*sizeof(__m512i)));
+        __m512i input5 = _mm512_loadu_si512((const __m512i *)(str + i + 4*sizeof(__m512i)));
+        __m512i input6 = _mm512_loadu_si512((const __m512i *)(str + i + 5*sizeof(__m512i)));
+        __m512i input7 = _mm512_loadu_si512((const __m512i *)(str + i + 6*sizeof(__m512i)));
+        __m512i input8 = _mm512_loadu_si512((const __m512i *)(str + i + 7*sizeof(__m512i)));
+
+        uint64_t continuation_bitmask[8];
+        continuation_bitmask[0] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input1, continuation));
+        continuation_bitmask[1] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input2, continuation));
+        continuation_bitmask[2] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input3, continuation));
+        continuation_bitmask[3] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input4, continuation));
+        continuation_bitmask[4] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input5, continuation));
+        continuation_bitmask[5] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input6, continuation));
+        continuation_bitmask[6] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input7, continuation));
+        continuation_bitmask[7] = static_cast<uint64_t>(_mm512_cmple_epi8_mask(input8, continuation));
+
+        __m512i mask_register = _mm512_loadu_si512(reinterpret_cast<const __m512i*>(continuation_bitmask));
+        #ifdef _MSC_VER
+          unrolled_popcount = _mm512_add_epi64(unrolled_popcount, _mm512_popcnt_epi64(mask_register));
+        #else
+          unrolled_popcount += _mm512_popcnt_epi64(mask_register); // Using += is slightly faster than using __mm512_add_epi64 on Linux
```

Comment:
I recommend against doing that. Your code is correct but its lack of portability is harmful. There is no good reason why this should not have the exact same speed: `unrolled_popcount = _mm512_add_epi64(unrolled_popcount, _mm512_popcnt_epi64(mask_register));`. If it differs by a lot, let us investigate.

Reply:
Taking a break for the night, but it is probably worth investigating: 

For += operator, I get: 
   0.413 ins/byte,    0.230 cycle/byte,   13.624 GB/s (1.3 %),     3.137 GHz,    1.794 ins/cycle 
   1.199 ins/char,    0.668 cycle/char,    4.696 Gc/s (1.3 %)     2.90 byte/char 

   for __mm512_add_epi_64, I get:
I get : 
   0.425 ins/byte,    0.243 cycle/byte,   12.886 GB/s (1.5 %),     3.134 GHz,    1.747 ins/cycle 
   1.233 ins/char,    0.705 cycle/char,    4.442 Gc/s (1.5 %)     2.90 byte/char

Reply:
...or maybe not. 

Turns out I was benchmarking with the debug version. I dunno what the compiler was doing under the hood but 
the release version is very similar in speed.

Reply:
See my comments below, I don't think it is should matter.